### ISSUE TITLE
Inconsistent is_on_wall() reporting behavior

### ISSUE THREAD

**Issue by SpicoTheSpiritGuy** [godotengine/godot#76756]
### Godot version

4.0.2stable

### System information

Windows 10, Intel Celeron N3060, Intel HD Graphics (Unknown)

### Issue description

the is_on_wall() reporting is inconsistent depending on what buttons are pressed, released, or what side the collision is happening at. 
On the ground the behavior is as intended (is on wall outputs true if you are only going towards a wall and not doing so when letting go), but mid air there are some inconsistencies:
- If you walk towards a wall to be on your left then jump without going towards a direction, the output will always be true. 
- If you do the same but towards a right wall however the output will always be false, though there happens to be a single frame period of time where the output is true. 
--- If you go towards the right with a wall on your right, the output will be stuck on true once again when you stop going towards the right.

### Steps to reproduce

1. Either approach a wall from any side without touching the ground, or become airborn next to a wall if you are grounded by the time you approach it.

### Minimal reproduction project

[Bug report.zip](https://github.com/godotengine/godot/files/11408816/Bug.report.zip)

**Comment by SpicoTheSpiritGuy** [comment]
The issue still appears to be present in Godot 4.0.3

**Comment by Calinou** [comment]
Can you reproduce this if you increase `safe_margin` on the CharacterBody2D? Try values such as 0.5, 1, 2, 4.

**Comment by SpicoTheSpiritGuy** [comment]
> Can you reproduce this if you increase `safe_margin` on the CharacterBody2D? Try values such as 0.2, 0.5, 1, 2, 4.

yes, it is reproducable with different safe margin values, and the behavior is the same between all of them.

**Comment by HVukman** [comment]
I saw the same issue. Seems to self-correct once you go right again.

**Comment by n500dev** [comment]
This problem persists in version 4.4 stable and may have even gotten worse.

Some days, when I run the game, the `CharacterBody2D` checks `is_on_wall_only()` correctly for both the left and right walls. On other days, **without changing anything in the code**, the `CharacterBody2D` only checks `is_on_wall_only()` correctly when the character jumps into contact with a wall on its left side, and fails when jumping while touching a wall on its right side.
On some occasions, the `CharacterBody2D` stops working correctly in the middle of the playthrough.

I tried adjusting `safe_margin` to 0.2, 0.5, 1, 2, or 4 as Calinou suggested. All it did was change which side the `CharacterBody2D` detects correctly or cause it to detect neither side.


I also tried printing something inside `physics_process` while jumping and touching a wall, and I got these results:

```
The character is jumping with the left side touching a wall:

Frame no. 6523: is_on_wall_only false   #(it's false because the character is also touching the ground at this frame)
Frame no. 6527: is_on_wall_only true    #(it detects correctly that the character is touching the wall while jumping)
Frame no. 6533: is_on_wall_only true    #(it keeps returning true because the character is only moving up, without any other input)
Frame no. 6538: is_on_wall_only true
Frame no. 6544: is_on_wall_only true
Frame no. 6555: is_on_wall_only true
[...] and so on
```

Versus

```
The character is jumping with the right side touching a wall:

Frame no. 7760: is_on_wall_only false   #(it's false because the character is also touching the ground at this frame)
Frame no. 7765: is_on_wall_only true    #(it detects correctly that the character is touching the wall while jumping)
Frame no. 7769: is_on_wall_only false   #(something happened without any input from me, and from now on, it will only return false)
Frame no. 7775: is_on_wall_only false
Frame no. 7781: is_on_wall_only false
Frame no. 7787: is_on_wall_only false
[...] and so on
```

In my case, it only happens if the `CharacterBody2D` jumps without pressing any direction key.

**Comment by paomian** [comment]
This problem persists in version 4.5 beta 2.
I use TileMapLayer build my game env.

```
14084 player pos: (170.4252, -270.2759) is on wall: true
14086 player pos: (170.4252, -267.1509) is on wall: true
14088 player pos: (170.4252, -263.9987) is on wall: true
14090 player pos: (170.4252, -260.8192) is on wall: true
14092 player pos: (170.4252, -257.6126) is on wall: true
14094 player pos: (170.4252, -254.3787) is on wall: true
14096 player pos: (170.4252, -251.1176) is on wall: false
14096 wall_sliding -> fall
14098 player pos: (170.4252, -251.1176) is on wall: false
14100 player pos: (170.4252, -247.5843) is on wall: false
14102 player pos: (170.4252, -243.7787) is on wall: false
14104 player pos: (170.4252, -239.7009) is on wall: false
14106 player pos: (170.4252, -235.3509) is on wall: false
14108 player pos: (170.4252, -230.7287) is on wall: false
14110 player pos: (170.4252, -225.8343) is on wall: false
14112 player pos: (170.4252, -220.6676) is on wall: false
14114 player pos: (170.4252, -215.2287) is on wall: true
14114 fall -> wall_sliding
14116 player pos: (170.4252, -215.2287) is on wall: true
14118 player pos: (170.4252, -214.1137) is on wall: true
14120 player pos: (170.4252, -212.9715) is on wall: true
14122 player pos: (170.4252, -211.802) is on wall: true
14124 player pos: (170.4252, -210.6054) is on wall: true
14126 player pos: (170.4252, -209.3815) is on wall: true
14128 player pos: (170.4252, -208.1304) is on wall: true
14130 player pos: (170.4252, -206.852) is on wall: true
14132 player pos: (170.4252, -205.5465) is on wall: true
14134 player pos: (170.4252, -204.2137) is on wall: true
```

I don't have any input on the way down.

**Comment by HalinSky** [comment]
I have encountered this issue in 4.4 while CharacterBody2D is jumping upwards when already touching the wall on the ground itself. 
Both is_on_wall() and is_on_wall_only() return false for me at all times in this situation.

**Comment by aladin20c** [comment]
same issue Godot 4.4 Mac M1
Hi, I m encountering the same problem. for my problem, for charcterBody2d. if is_on_wall_only is true it can go up or down by using arrows otherwise it falls. 
when I stack two square static2d bodies and start climbing. when I reach the border that separates them is_on_wall_only() suddenly returns false and I fall. 

the same behavior is reproduces with is_on_wall() and with different safe margins I tried from 0.08, 0.083, 0.09, 0.1, 0.2, 0.3, 0.4, 0.5, 1, 2, 3, 4.

**Comment by SpicoTheSpiritGuy** [comment]
Tested this after porting the minimum reproduction project attached here from 4.0 to 4.4.1, and managed to find two workarounds. Doing either one of these "fixes" the problem:

- Setting the floor_block_on_wall of the characterbody2d to false with no other adjustments fixes the collisions mid air, but godot gets confused when you jump next to a wall, is_on_wall() becoming true after you land until you walk into the wall again (it also quickly flickers from false to true)
- Setting the safe_margin to 0.01 pixels of the characterbody2d with no other adjustments fixes it completely. Any lower, it breaks on the right wall, and any higher, it breaks on the left wall. 2 also seems to work but that has some more problems, like is_on_wall() flickering from true to false every frame when you jump and land next to a wall. I found 0.01 works the best and feels the most consistent between both walls

The fixes seem to work on the reproduction project, with the latter being much more effective. None of these two fix it completely though, it is still inconsistent in some ways

**Comment by xuhuisheng** [comment]
@SpicoTheSpiritGuy 
It is wired that I changed safe margin from 0.08 to 0.09, then right wall slide will return is_on_wall() == true, consistently. 

I need keep digging.

**Comment by xuhuisheng** [comment]
@robotlovesyou 
Could you create a demo for testing? Or you can using my patch to confirm whether it can fix your issue.

#110801
